File: DnDSpellBook/DnDSpellBook/app/Controllers/characterController.js
app.controller("characterController", ["$scope", "$http", function ($scope, $http) {

    $scope.characters = [];

    $scope.charactersSpells = [];

    $scope.selectedCharacter = {};

    $scope.selectedSpell = [];

    $scope.cardDetail = [];
    
    //gets list of characters
    $scope.getCharacters = function () {
        $http.get(`/api/characters`)
            .then(function (result) {
                $scope.characters = result.data;
                console.log("characters", $scope.characters);
                console.log(result);
            });
    };

    $scope.getCharacters();

    //get character spells
    $scope.getCharacterSpells = function (selectedCharacterId) {
        console.log(selectedCharacterId);
        $http.get(`/api/characters/spells/${selectedCharacterId}`)
            .then(function (result) {
                $scope.charactersSpells = result.data;
                console.log("spells returned", $scope.charactersSpells);
                console.log(result);
            });
    };

    //gets spell detail
    $scope.spellDetail = function (characterSpell, $index) {
        spellurl = encodeURIComponent(characterSpell.url);
        console.log("url passed in", spellurl);
        if ($scope.cardDetail[$index] == true) {
            $scope.cardDetail[$index] = false;
        } else {
            $scope.cardDetail[$index] = true;
        }
        $http.get(`/api/spells/?spellurl=${spellurl}`)
           .then(function (result) {
               $scope.charactersSpells[$index].spelldetail = result.data;
               //console.log("spelldetail", characterSpell.spelldetail);
               console.log(result);
           });

    };

    //delete spell
    $scope.deleteSpell = function (spellName, characterId) {
        console.log("Spell to be deleted", spellName);
        console.log("characterID", characterId);
        $http.delete(`/api/character/spell/delete/${spellName}/${characterId}`)
            .then(function () {
                $scope.getCharacterSpells(characterId);
            });
    };

    //deletes character
    $scope.deleteCharacter = function (selectedCharacterId) {
        console.log("SelectedcharId", selectedCharacterId);
        $http.delete(`/api/character/delete/${selectedCharacterId}`)
    };
}]);

File: DnDSpellBook/DnDSpellBook/Controllers/CharacterController.cs
using DnDSpellBook.DAL;
using DnDSpellBook.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Web.Http;

namespace DnDSpellBook.Controllers
{
    public class CharacterController : ApiController
    {
        readonly ICharacterRepository _characterRepository;

        public CharacterController(ICharacterRepository CharacterRepository)
        {
            _characterRepository = CharacterRepository;
        }

        [HttpGet, Route("api/characters")]
        public IEnumerable<Character> GetCharacters()
        {
            return _characterRepository.GetCharacters();
        }

        //gets spells list of selected character
        [HttpGet, Route("api/characters/spells/{selectedCharacterId}")]
        public IEnumerable<Spell> GetSpells(int selectedCharacterId)
        {
            return _characterRepository.GetCharactersSpells(selectedCharacterId);
        }

        [HttpDelete, Route("api/character/delete/{selectedCharacterId}")]
        public void DeleteCharacter(int selectedCharacterId)
        {
            _characterRepository.DeleteCharacter(selectedCharacterId);
        }
        
        [HttpDelete, Route("api/character/spell/delete/{spellName}/{characterId}")]
        public void DeleteSpell(string spellName, int characterId)
        {
            _characterRepository.DeleteSpell(spellName, characterId);
        }
        //deletes spell from list

        //deletes character from db
    }
}
